### PR TITLE
feat: add audioclip into the flow

### PR DESCRIPTION
**Pull request by nan-wang**
todo:
- [x] Add AudioCLIP
- [x] Fix the AudioCLIP Hub executors https://github.com/jina-ai/executors/pull/266
- [x] Implement a Ranker to mix the results from both image and audio sources
- [x] Write the README
- [x] Segment audio tracks into chunks
---

As for the data, I suggest to use the following, **please don't add these two data into the repo. They are quite large and it is better to write a script to download them**

```bash

# video of different animal sounds with pictures
youtube-dl --recode-video mp4 -o 0GhwfyDcAoM.mp4 https://www.youtube.com/watch\?v\=0GhwfyDcAoM

# video of different housing sounds with pictures
youtube-dl --recode-video mp4 -o uTNL3WXEexA.mp4 https://www.youtube.com/watch\?v\=uTNL3WXEexA
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,32 +1,88 @@
...
+| ------------- | ------------- |
+| Learnings | How to search through both image frames and audio of a video. |
+| Used for indexing | Video Files. |
+| Used for querying | Text Query (e.g. "girl studying engineering") |
+| Dataset used | Choose your own videos |
+| Model used | [AudioCLIP](https://github.com/AndreyGuzhov/AudioCLIP) |
+
+In this example, Jina is used to implement a video search system.
+Videos can be indexed and afterwards searched by text inputs. 
+jina searches both the image frames and the audio of the video and returns
```
Comment (cristianmtr) [inline README.md:25]:
```suggestion
Jina searches both the image frames and the audio of the video and returns
```

File: README.md [hunk 2]
```diff
@@ -1,32 +1,88 @@
...
+_____
+
+## 🐍 Build the app with Python
+
+These instructions explain how to build the example yourself and deploy it with Python.
+
+
+### 🗝️ Requirements
+
+1. You have a working Python 3.7 or 3.8 environment and a installation of [Docker](https://docs.docker.com/get-docker/), assure that you set enough memory resources(more than 6GB) to the docker. You can set it in settings/resources/advanced in your Docker.
```
Comment (cristianmtr) [inline README.md:37]:
```suggestion
1. You have a working Python 3.7 or 3.8 environment and a installation of [Docker](https://docs.docker.com/get-docker/). Ensure that you set enough memory resources(more than 6GB) to docker. You can set it in settings/resources/advanced in Docker.
```

File: README.md [hunk 3]
```diff
@@ -1,32 +1,88 @@
...
+2. We recommend creating a [new Python virtual environment](https://docs.python.org/3/tutorial/venv.html) to have a clean installation of Jina and prevent dependency conflicts.   
+3. You have at least 5 GB of free space on your hard drive. 
+
+### 👾 Step 1. Clone the repo and install Jina
+
+Begin by cloning the repo, so you can get the required files and datasets. (If you already have the examples repository on your machine make sure to fetch the most recent version)
```
Comment (cristianmtr) [inline README.md:43]:
```suggestion
Begin by cloning the repo, so you can get the required files and datasets.
```

File: index-flow.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    volumes: $WORKSPACE_MOUNT
+    needs: ['image_encoder']
+  - name: audio_filter
+    uses: FilterModality
+    uses_with:
+      modality: 'audio'
+    py_modules:
+      - executors.py
+    needs: frame_extractor
+  - name: audio_encoder
```
Comment (nan-wang) [inline index-flow.yml:36]:
```suggestion
  - name: audio_segmenter
     uses: AudioSegmenter
     py_modules:
       - executors.py
  - name: audio_encoder
```

File: index-flow.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    uses: jinahub+docker://AudioCLIPEncoder/v0.4
+    uses_with:
+      traversal_paths: ['c',]
```
Comment (nan-wang) [inline index-flow.yml:39]:
```suggestion
      traversal_paths: ['cc',]
```

File: executors.py [hunk 6]
```diff
@@ -20,97 +20,23 @@
...
-
-    def _get_timestamp_from_filename(self, uri):
-        return os.path.basename(uri).split('.')[0]
+            chunks = filter(lambda d: d.modality == self.modality, doc.chunks)
+            doc.chunks = chunks
+        return docs
 
 
-class SimpleRanker(Executor):
+class MixRanker(Executor):
```
Comment (nan-wang) [inline executors.py:39]:
```suggestion
class AudioSegmenter(Executor):
    def __init__(self, chunk_duration: int = 10, chunk_strip: int = 1, traversal_paths = ('r', ), *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.chunk_duration = chunk_duration  # seconds
        self.strip = chunk_strip

    @requests(on=['/search', '/index'])
    def segment(
        self, docs: Optional[DocumentArray] = None, parameters: dict = {}, **kwargs
    ):
        if not docs:
            return
                    traversal_paths = parameters.get('traversal_paths', self.traversal_paths)

        for idx, doc in enumerate(docs.traverse_flat(traversal_paths)):
            sample_rate = doc.tags['sample_rate'] = sample_rate
            chunk_size = int(self.chunk_duration * sample_rate)
            strip = parameters.get('chunk_strip', self.strip)
            strip_size = int(strip * sample_rate)
            num_chunks = max(1, int((doc.blob.shape[0] - chunk_size) / strip_size))
            for chunk_id in range(num_chunks):
                beg = chunk_id * strip_size
                end = beg + chunk_size
                if beg > doc.blob.shape[0]:
                    break
                doc.chunks.append(
                    Document(
                        blob=doc.blob[beg:end],
                        offset=idx,
                        location=[beg, end],
                        tags=doc.tags,
                    )
                )
 
class MixRanker(Executor):
```

**Comment by jakob1111996** [comment]
For some reason, I can not add @nan-wang as reviewer.

File: search-flow.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    uses: jinahub://SimpleIndexer/v0.4
+    uses_with:
+      match_args:
+        limit: $TOP_K
+        traversal_ldarray: ['r',]
+        traversal_rdarray: ['c',]
+    uses_metas:
+      workspace: $JINA_WORKSPACE/audio_indexer
+    volumes: $WORKSPACE_MOUNT
+    needs: ['image_indexer']
```
Comment (nan-wang) [inline search-flow.yml:35]: The simpleindexer will override the matches. Plus, we need to `MatchMerger` to merge the matches from the two pathways.

Reference:
https://github.com/jina-ai/jina/blob/master/jina/types/arrays/neural_ops.py#L124
```suggestion
    needs: ['text_encoder']
```

File: index-flow.yml [hunk 8]
```diff
@@ -0,0 +1,57 @@
+jtype: Flow
+version: '1'
+with:
+  protocol: 'http'
+  cors: True
+  port_expose: $JINA_PORT
+executors:
+  - name: frame_extractor
+    uses: jinahub://VideoLoader/v0.1
```
Comment (nan-wang) [inline index-flow.yml:9]:
```suggestion
    uses: jinahub://VideoLoader/v0.1
    uses_with:
      modality_list:
        - 'image'
        - 'audio'
```

File: index-flow.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    uses: FilterModality
+    uses_with:
+      modality: 'audio'
+    py_modules:
+      - executors.py
+    needs: frame_extractor
+  - name: audio_segmenter
+    uses: AudioSegmenter
+    uses_with:
+      traversal_paths: ['c',]
```
Comment (nan-wang) [inline index-flow.yml:39]: reduce the overlaps between the chunks.
```suggestion
      traversal_paths: ['c',]
      chunk_strip: 10
```

**Comment by nan-wang** [comment]
Another thing is that I find the score from the audios is always lower than that from the images.

**Pull request merged**